### PR TITLE
Added Ordnance Survey providers

### PR DESCRIPTION
**Pull request by aw-west-defra**
OS Maps API  
[get a free key](https://osdatahub.os.uk/)  
[OS Maps documentation](https://osdatahub.os.uk/docs/wmts/technicalSpecification)   
[OS attribution](https://www.ordnancesurvey.co.uk/business-government/licensing-agreements/copyright-acknowledgements)

### CODE REVIEW

File: provider_sources/xyzservices-providers.json [hunk 1]
```diff
@@ -373,5 +373,98 @@
             "max_zoom": 12,
             "name": "MapTiler.Terrain"
         }
+    },
+    "OrdnanceSurvey": {
+        "Road_3857": {
+            "url": "https://api.os.uk/maps/raster/v1/zxy/Road_3857/{z}/{x}/{y}.png?key={key}",
+            "html_attribution": "Contains OS data &copy Crown copyright and database right {year}",
+            "attribution": "Contains OS data (C) Crown copyright and database right {year}",
+            "key": "<A valid OS MapsAPI Key. Get a free key here - https://osdatahub.os.uk/>",
```
Comment (martinfleis) [inline provider_sources/xyzservices-providers.json:382]:
```suggestion
            "key": "<insert your valid OS MapsAPI Key. Get a free key here - https://osdatahub.os.uk/>",
```

these placeholders need to start with `"<insert your"` as we use this string to check if a token is required or not (maybe a bit clumsily but that's how it is done now).

File: provider_sources/xyzservices-providers.json [hunk 2]
```diff
@@ -373,5 +373,98 @@
...
+            "crs": "EPSG:3857",
+            "min_zoom": 7,
+            "max_zoom": 16,
+            "max_zoom_premium": 20,
+            "bounds": [[-841264.4659119569, 6405988.404624734], [404617.877362011, 8733428.658192052]],
```
Comment (martinfleis) [inline provider_sources/xyzservices-providers.json:387]: I suppose that bounds for the default 3857 CRS should actually be in lat lon so you can directly pass them to leaflet and similar (https://leafletjs.com/reference.html#latlngbounds) as documented in our [contributing guide](https://xyzservices.readthedocs.io/en/stable/contributing.html).

**Review by martinfleis — COMMENTED** [review]
Thanks! Two minor notes left in the code.

I'll just need to remind myself to make a follow-up PR adding a test once I store the token in the repo environment variables.

File: provider_sources/xyzservices-providers.json (hunk 2 above)
Comment (martinfleis) [reply provider_sources/xyzservices-providers.json:387]: The bounds shall be in latitude longitude coordinates (in degrees), not in the unit of 3857 projection. That is what I was trying to say. The 3857 is the default for the XYZ standard and applications assume 3857 of tiles but bounds in lat long.

Comment (aw-west-defra) [reply provider_sources/xyzservices-providers.json:387]: I made a mistake in projecting from BNG.
BNG = 0,0,7e5,13e5
3857 = -9.496386752661158, 49.76680723514262, 3.634744234622926, 61.465188864662096

Should I be including these bounds?
They're not explicitly stated for 3857, only BNG.

Comment (martinfleis) [reply provider_sources/xyzservices-providers.json:387]: Yes, include `49.76680723514262, -9.496386752661158, 61.465188864662096, 3.634744234622926` for those tiles that are in 3857. Maybe with a bit less decimals :). 

Bounds for those in 27700 can stay as they are like we have in cases of other custom projections.

File: provider_sources/xyzservices-providers.json [hunk 3]
```diff
@@ -373,5 +373,98 @@
             "max_zoom": 12,
             "name": "MapTiler.Terrain"
         }
+    },
+    "OrdnanceSurvey": {
+        "Road_3857": {
+            "url": "https://api.os.uk/maps/raster/v1/zxy/Road_3857/{z}/{x}/{y}.png?key={key}",
+            "html_attribution": "Contains OS data &copy Crown copyright and database right {year}",
+            "attribution": "Contains OS data (C) Crown copyright and database right {year}",
```
Comment (martinfleis) [inline provider_sources/xyzservices-providers.json:381]: I've now also noticed the `{year}` placeholder here. We currently don't support placeholders outside of the `url` parameter, so you will need to spell it out here. The attribution here shall always have the current year? We may need to figure out some autoupdating logic within https://github.com/geopandas/xyzservices/blob/main/provider_sources/_compress_providers.py 

I can probably take care of it within the follow-up PR alognside tests.

File: provider_sources/xyzservices-providers.json (hunk 2 above)
Comment (aw-west-defra) [reply provider_sources/xyzservices-providers.json]: Rounded to 9sf like SwissFederalGeoportal:  `49.766807, -9.496386, 61.465189, 3.634745`

**Comment by martinfleis** [comment]
All good now. Thanks!

**Pull request merged**